### PR TITLE
Jenkinsfile change to support branching

### PR DESCRIPTION
**Pull request by prudhvigodithi**
Signed-off-by: prudhvigodithi <pgodithi@amazon.com>

### Description
Jenkinsfile change to support branching, starting with daily used jenkins jobs of OpenSearch and OpenSearch dashboard.

### Issues Resolved
Part of: https://github.com/opensearch-project/opensearch-build/issues/2549
Related: https://github.com/opensearch-project/opensearch-build/issues/2558

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by gaiksaya** [comment]
Hi @prudhvigodithi, can you check this issue? https://github.com/opensearch-project/opensearch-build/issues/1606
We can put this functionality into a common pipeline which then can be consumed by all JenkinsFiles

**Comment by peterzhuamazon** [comment]
There are too many stash unstash and I wonder is there a way to load a common lib like @gaiksaya suggested.

**Comment by prudhvigodithi** [comment]
The problem I see is git checkout, yes can be added to shared library, still that shared library function has to be called in jenkinsfile, but I dont see a way we can escape the unstash line, as jenkins expects for every stage to have that unstash method so that it can re-use those cloned files, will do some more research here.
@peterzhuamazon @gaiksaya 
Thank you

**Comment by prudhvigodithi** [comment]
Even to load the shared library we need to add a stage/line to checkout and stash, I dont see a huge difference here to add 2 line in the jenkinsfile itself that does the git checkout and stash the files, also to unstash I dont see any other way to declare globally other then to have `unstash` part of each stage.

File: jenkins/opensearch-dashboards/bwc-test.jenkinsfile [hunk 1]
```diff
@@ -20,20 +22,40 @@ pipeline {
             description: 'The build manifest URL, e.g. https://ci.opensearch.org/ci/dbc/distribution-build-opensearch-dashboards/2.0.0/98/linux/x64/builds/opensearch-dashboards/manifest.yml.',
             trim: true
         )
+        string(
+            name: 'BUILD_BRANCH',
+            description: "Build repo branch to use.",
+            defaultValue: "main",
+            trim: true
+        )
```
Comment (dblock) [inline jenkins/opensearch-dashboards/bwc-test.jenkinsfile:30]: This feels very wrong. This job itself lives in this repo in a branch. If I am running the job out branch X, I would not expect the job to check out build tools from branch Y.

**Comment by dblock** [comment]
I think you want to model this after GitHub actions. The .yml jobs for GHA live in a branch (e.g. 1.2.3) and run from that branch (1.2.3) checking out source code by default from the same branch (1.2.3). This it's never a parameter to the job, the only time you care about which branch to use is when you import the job into Jenkins.

File: jenkins/opensearch-dashboards/bwc-test.jenkinsfile (hunk 1 above)
Comment (prudhvigodithi) [reply jenkins/opensearch-dashboards/bwc-test.jenkinsfile:30]: This is done for the fact Jenkinsfiles cannot be branched, they has to reside in `main` and jenkins job git SCM configuration uses single branch to pull the jenkinsfile, once jenkinsfile is pulled, depending on the branch passed as input using `BUILD_BRANCH`, then the checkout happens with right branch and all the stages are executed as per the flow.

**Comment by prudhvigodithi** [comment]
> I think you want to model this after GitHub actions. The .yml jobs for GHA live in a branch (e.g. 1.2.3) and run from that branch (1.2.3) checking out source code by default from the same branch (1.2.3). This it's never a parameter to the job, the only time you care about which branch to use is when you import the job into Jenkins.

Hey @dblock, assume the build repo is branched as `main`, `2.x` and `3.x`, now jenkinsfiles can only be on `main`, since the job configuration need to pull the jenkinsfiles from a fixed branch (which has be configured inside the job settings), now we have to build workflow code part of `2.x` (we are releasing `2.3.0`), the job clones the jenkinsfiles from `main` and switch the branch to `2.`x using the `BUILD_BRANCH` parameter, this way the workspace now has the right `2.x` release code to run the build and assemble.

**Comment by gaiksaya** [comment]
I think what is causing the confusion is job configuration (one time thing manually for now) is where we state from where will the related JenkinsFile be pulled up. 
See the snippet below:
image
Need to check if that `branch` is configurable via code.

**Comment by prudhvigodithi** [comment]
We can go with this PR until the [Branching Strategy Implementation](https://github.com/opensearch-project/opensearch-build/issues/2549) is finalized, I dont see any harm that would break the existing distribution functionality. This input value `BUILD_BRANCH` would be basic requirement to fetch the right branch code.
Since the release [2.3.0](https://github.com/opensearch-project/opensearch-build/issues/2447) is ongoing waiting to get a RC candidate for 2.3 before we can merge this.

**Comment by dblock** [comment]
I think this is the wrong implementation for an otherwise good idea. As an example, your `distribution-build.jenkinsfile` comes from `main`, but you're trying to specify `1.2.3` as a build branch. The problem is that it's the branch in _the same repo_, so your .jenkinsfile may or may not work with version `1.2.3` at all. 

I think you want to extract the python tooling into a new repo, then hard-code the branch to use from that repo in the manifests, not take a parameter in build. This would make it reproducible and configurable in 1 place.

**Comment by prudhvigodithi** [comment]
Hey @dblock I like the idea to separate the python tooling and pull the right tooling from manifest using the branch (or version), but its just not python tooling, we have some scripts here as well https://github.com/opensearch-project/opensearch-build/tree/main/scripts that needs to be focused for branching.

> The problem is that it's the branch in _the same repo_, so your .jenkinsfile may or may not work with version `1.2.3` at all.

@dblock Jenkinsfile cannot be branched, hence the are only part of `main` , shared libraries are already being separated out and will be imported by version, jenkinsfile will switch the branch and execute the stages as per release. So ideally same .jenkinsfile will be used for 1.x and 2.x, i mean if there is a drastic change in jenkinsfile, even with separation there are chances pipeline fails for past releases as well ;).

How about we get this merged, observe with `BUILD_BRANCH` input and then if required consider for separating the python tooling or even other scripts if required?

If no then I will close this PR and start focusing on python workflow separation and consider branching(or versioning) on those separated repos. :)  Once done then some versioning has to be done for those scripts as well https://github.com/opensearch-project/opensearch-build/tree/main/scripts 
@bbarani @peterzhuamazon @gaiksaya 
Thank you

**Comment by dblock** [comment]
I recommend closing and focusing on branching those new repos, because this is a step in the wrong direction.

**Comment by prudhvigodithi** [comment]
Closing this PR. More details https://github.com/opensearch-project/opensearch-build/issues/2549#issuecomment-1244334846, https://github.com/opensearch-project/opensearch-build/issues/2549#issuecomment-1245859146
Thank you

**Pull request closed without merging**